network-libp2p/src/autonat.rs

```diff
@@ -0,0 +1,97 @@
+use std::collections::{HashMap, HashSet};
+
+use libp2p::Multiaddr;
+
+#[derive(Clone, Debug, Default, PartialEq)]
+pub(crate) enum NatStatus {
+    Public,
+    Private,
+    #[default]
+    Unknown,
+}
+
+#[derive(Default)]
+pub(crate) struct NatState {
+    confirmed_addresses: HashSet<Multiaddr>,
+    listen_address_status: HashMap<Multiaddr, NatStatus>,
+    status: NatStatus,
+}
+
+impl NatState {
+    pub fn add_listen_address(&mut self, address: Multiaddr) {
+        self.listen_address_status
+            .insert(address, NatStatus::Unknown);
+    }
+
+    pub fn remove_listen_address(&mut self, address: &Multiaddr) {
+        self.listen_address_status.remove(address);
+        self.confirmed_addresses.remove(address);
+        self.update_state();
+    }
+
+    pub fn set_listen_address_nat_status(&mut self, address: Multiaddr, nat_status: NatStatus) {
+        if let Some(status) = self.listen_address_status.get_mut(&address) {
+            *status = nat_status.clone();
+
+            if nat_status == NatStatus::Public {
+                self.confirmed_addresses.insert(address);
+            } else {
+                self.confirmed_addresses.remove(&address);
+            }
```

Comment:
Swap around to avoid the `.clone()`.
```suggestion
            if nat_status == NatStatus::Public {
                self.confirmed_addresses.insert(address);
            } else {
                self.confirmed_addresses.remove(&address);
            }

            *status = nat_status;
```

Reply:
Swap not necessary since it's now a `Copy` 👍